File: device/metal/causal/native/adjustment.m
#include "adjustment.h"
#include "causal.h"
#include "../internal/bridge/core_private.h"

#include <Foundation/Foundation.h>
#include <Metal/Metal.h>

int metal_dispatch_backdoor_adjustment(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef conditionalRef,
    MetalBufferRef marginalRef,
    MetalBufferRef outRef,
    uint32_t xCount,
    uint32_t zCount,
    uint32_t yCount,
    uint64_t completionToken,
    MetalStatus* status
) {
    if (conditionalRef == NULL || marginalRef == NULL || outRef == NULL) {
        metal_causal_status_set(status, -2, "nil Metal buffer");
        return -2;
    }

    return metal_causal_named_dispatch(
        contextRef, elementDType, "backdoor_adjustment", (NSUInteger)xCount * yCount,
        completionToken, status,
        ^(id<MTLComputeCommandEncoder> encoder) {
            [encoder setBuffer:(__bridge id<MTLBuffer>)conditionalRef offset:0 atIndex:0];
            [encoder setBuffer:(__bridge id<MTLBuffer>)marginalRef offset:0 atIndex:1];
            [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:2];
            [encoder setBytes:&xCount length:sizeof(xCount) atIndex:3];
            [encoder setBytes:&zCount length:sizeof(zCount) atIndex:4];
            [encoder setBytes:&yCount length:sizeof(yCount) atIndex:5];
        }
    );
}

int metal_dispatch_frontdoor_adjustment(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef mediatorRef,
    MetalBufferRef outcomeRef,
    MetalBufferRef marginalRef,
    MetalBufferRef outRef,
    uint32_t xCount,
    uint32_t mCount,
    uint32_t yCount,
    uint64_t completionToken,
    MetalStatus* status
) {
    if (mediatorRef == NULL || outcomeRef == NULL || marginalRef == NULL || outRef == NULL) {
        metal_causal_status_set(status, -2, "nil Metal buffer");
        return -2;
    }

    return metal_causal_named_dispatch(
        contextRef, elementDType, "frontdoor_adjustment", (NSUInteger)xCount * yCount,
        completionToken, status,
        ^(id<MTLComputeCommandEncoder> encoder) {
            [encoder setBuffer:(__bridge id<MTLBuffer>)mediatorRef offset:0 atIndex:0];
            [encoder setBuffer:(__bridge id<MTLBuffer>)outcomeRef offset:0 atIndex:1];
            [encoder setBuffer:(__bridge id<MTLBuffer>)marginalRef offset:0 atIndex:2];
            [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:3];
            [encoder setBytes:&xCount length:sizeof(xCount) atIndex:4];
            [encoder setBytes:&mCount length:sizeof(mCount) atIndex:5];
            [encoder setBytes:&yCount length:sizeof(yCount) atIndex:6];
        }
    );
}

File: device/cuda/causal/native/adjustment.c
#include "adjustment.h"
#include "causal_dispatch.h"
#include "../internal/bridge/core_private.h"

int cuda_dispatch_backdoor_adjustment(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef conditionalRef,
    CUDABufferRef marginalRef,
    CUDABufferRef outRef,
    uint32_t xCount,
    uint32_t zCount,
    uint32_t yCount,
    uint64_t completionToken,
    CUDAStatus* status
) {
    if (conditionalRef == NULL || marginalRef == NULL || outRef == NULL) {
        cuda_causal_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    void* conditionalPtr = cuda_buffer_device_ptr(conditionalRef);
    void* marginalPtr = cuda_buffer_device_ptr(marginalRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&conditionalPtr, &marginalPtr, &outPtr, &xCount, &zCount, &yCount};

    return cuda_causal_named_launch(
        contextRef,
        elementDType,
        "backdoor_adjustment",
        xCount * yCount,
        args,
        sizeof(args),
        completionToken,
        status
    );
}

int cuda_dispatch_frontdoor_adjustment(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef mediatorRef,
    CUDABufferRef outcomeRef,
    CUDABufferRef marginalRef,
    CUDABufferRef outRef,
    uint32_t xCount,
    uint32_t mCount,
    uint32_t yCount,
    uint64_t completionToken,
    CUDAStatus* status
) {
    if (mediatorRef == NULL || outcomeRef == NULL || marginalRef == NULL || outRef == NULL) {
        cuda_causal_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    void* mediatorPtr = cuda_buffer_device_ptr(mediatorRef);
    void* outcomePtr = cuda_buffer_device_ptr(outcomeRef);
    void* marginalPtr = cuda_buffer_device_ptr(marginalRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&mediatorPtr, &outcomePtr, &marginalPtr, &outPtr, &xCount, &mCount, &yCount};

    return cuda_causal_named_launch(
        contextRef,
        elementDType,
        "frontdoor_adjustment",
        xCount * yCount,
        args,
        sizeof(args),
        completionToken,
        status
    );
}
